File: spring.py
# uncomment the next line if running in a notebook
# %matplotlib inline
import numpy as np
import matplotlib.pyplot as plt
import math
import time

start = time.time()

# mass, spring constant, initial position and velocity
m = 1
k = 1
x = 0
v = 1
x_verlet = x

# simulation time, timestep and time
t_max = 100
dt = 0.001
t_array = np.arange(0, t_max, dt)
xmin1 = -v*dt + x

# initialise empty lists to record trajectories
x_list = []
v_list = []
x_true_list = []
x_verlet_list = []


# Euler integration
for t in t_array:

    # append current state to trajectories
    x_list.append(x)
    v_list.append(v)
    x_true_list.append(math.sin(t))
    x_verlet_list.append(x_verlet)


    # calculate new position and velocity
    a = -k * x / m
    x = x + dt * v
    v = v + dt * a
    x_verlet_old = x_verlet
    a_verlet = -k * x_verlet / m
    x_verlet = 2*(x_verlet) - xmin1 + (dt**2)* a_verlet
    xmin1 = x_verlet_old

# convert trajectory lists into arrays, so they can be sliced (useful for Assignment 2)
x_array = np.array(x_list)
v_array = np.array(v_list)
x_true_array = np.array(x_true_list)
x_verlet_array = np.array(x_verlet_list)


# #£ plot the position-time graph
# plt.figure(1)
# plt.clf()
# plt.xlabel('time (s)')
# plt.grid()
# #£ plt.plot(t_array, x_array, label='x (m)')
# #£ plt.plot(t_array, v_array, label='v (m/s)')
# plt.plot(t_array, x_true_array, label='x_true (m)')
# plt.plot(t_array, x_verlet_array, label='x_verlet (m)')
# plt.legend()
# plt.show()

# printing to a file
f = open('trajectories_python.txt', 'w')

txt = ''
for i in np.arange(0, len(t_array)):
    txt += "{} {} {}\n".format(t_array[i], x_array[i], v_array[i])
f.write(txt)
f.close()

end = time.time()
print(end - start)

File: spring.cpp
#include <iostream>
#include <fstream>
#include <vector>

using namespace std;
using namespace std::chrono;


void euler_calc(double m, double k, double x, double v, double t_max, double dt){
  double t, a;
  vector<double> t_list, x_list, v_list;

  // mass, spring constant, initial position and velocity
  // m = 1;
  // k = 1;
  // x = 0;
  // v = 1;

  // simulation time and timestep
  // t_max = 100;
  // dt = 0.001;

  // Euler integration
  for (t = 0; t <= t_max; t = t + dt) {

    // append current state to trajectories
    t_list.push_back(t);
    x_list.push_back(x);
    v_list.push_back(v);

    // calculate new position and velocity
    a = -k * x / m;
    x = x + dt * v;
    v = v + dt * a;

  }
  // Write the trajectories to file
  ofstream fout;
  fout.open("euler_trajectories.txt");
  if (fout) { // file opened successfully
    for (int i = 0; i < t_list.size(); i = i + 1) {
      fout << t_list[i] << ' ' << x_list[i] << ' ' << v_list[i] << endl;
    }
  } else { // file did not open successfully
    cout << "Could not open trajectory file for writing" << endl;
  }
}

  //Verlet integration
void verlet_calc(double m, double k, double x, double v, double t_max, double dt){
  double t, a, xmin1, xold;
  vector<double> t_list, x_list, v_list ;
  xmin1 = -v*dt + x;
  

  for (t = 0; t <= t_max; t = t + dt) {
    xold = x;
    // append current state to trajectories
    t_list.push_back(t);
    x_list.push_back(x);
    v_list.push_back(v);

    // calculate new position and velocity
    
    a = -k * x / m;
    x = 2*(x) - xmin1 + (dt*dt)* a;
    xmin1 = xold;
  }




  // Write the trajectories to file
  ofstream fout;
  fout.open("verlet_trajectories.txt");
  if (fout) { // file opened successfully
    for (int i = 0; i < t_list.size(); i = i + 1) {
      fout << t_list[i] << ' ' << x_list[i] << ' ' << v_list[i] << endl;
    }
  } else { // file did not open successfully
    cout << "Could not open trajectory file for writing" << endl;
  }
}

int main() {

  // declare variables
  high_resolution_clock::time_point endtime, starttime = high_resolution_clock::now();
  microseconds duration;
  

  euler_calc(1, 1, 0, 1, 100, 0.001);
  verlet_calc(1, 1, 0, 1, 100, 0.001);
  
  // double m, k, x, v, t_max, dt, t, a;
  // vector<double> t_list, x_list, v_list;

  // // mass, spring constant, initial position and velocity
  // m = 1;
  // k = 1;
  // x = 0;
  // v = 1;

  // // simulation time and timestep
  // t_max = 100;
  // dt = 0.001;

  // // Euler integration
  // for (t = 0; t <= t_max; t = t + dt) {

  //   // append current state to trajectories
  //   t_list.push_back(t);
  //   x_list.push_back(x);
  //   v_list.push_back(v);

  //   // calculate new position and velocity
  //   a = -k * x / m;
  //   x = x + dt * v;
  //   v = v + dt * a;

  // }

  // //Verlet integration


  // // Write the trajectories to file
  // ofstream fout;
  // fout.open("trajectories.txt");
  // if (fout) { // file opened successfully
  //   for (int i = 0; i < t_list.size(); i = i + 1) {
  //     fout << t_list[i] << ' ' << x_list[i] << ' ' << v_list[i] << endl;
  //   }
  // } else { // file did not open successfully
  //   cout << "Could not open trajectory file for writing" << endl;
  // }

  endtime = high_resolution_clock::now();
  duration = duration_cast<microseconds>(endtime - starttime);

  cout << "Time taken by function: "
    << duration.count() << " microseconds" << endl;


  return 0;
  
  /* The file can be loaded and visualised in Python as follows:

  import numpy as np
  import matplotlib.pyplot as plt
  results = np.loadtxt('trajectories.txt')
  plt.figure(1)
  plt.clf()
  plt.xlabel('time (s)')
  plt.grid()
  plt.plot(results[:, 0], results[:, 1], label='x (m)')
  plt.plot(results[:, 0], results[:, 2], label='v (m/s)')
  plt.legend()
  plt.show()

  */
}
